sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -1353,14 +1236,131 @@ void set(
+      }
+      return input.apply(ParDo.of(new WriteFn<>(spec)));
+    }
+
+    private StaticValueProvider<String> generateStatement(List<SchemaUtil.FieldWithIndex> fields) {
+      return StaticValueProvider.of(
+          JdbcUtil.generateStatement(
+              getTable(),
+              fields.stream().map(FieldWithIndex::getField).collect(Collectors.toList())));
+    }
+
+    private List<SchemaUtil.FieldWithIndex> getFilteredFields(Schema schema) {
+      Schema tableSchema;
+
+      try (Connection connection = getDataSourceProviderFn().apply(null).getConnection();
+          PreparedStatement statement =
+              connection.prepareStatement((String.format("SELECT * FROM %s", getTable())))) {
+        tableSchema = SchemaUtil.toBeamSchema(statement.getMetaData());
+      } catch (SQLException e) {
+        throw new RuntimeException("Error while determining columns from table: " + getTable(), e);
+      }
+
+      if (tableSchema.getFieldCount() < schema.getFieldCount()) {
+        throw new RuntimeException("Input schema has more fields than actual table.");
+      }
 
-      return input.apply(ParDo.of(new WriteFn<>(this)));
+      // filter out missing fields from output table
+      List<Schema.Field> missingFields =
+          tableSchema.getFields().stream()
+              .filter(
+                  line ->
+                      schema.getFields().stream()
+                          .noneMatch(s -> s.getName().equalsIgnoreCase(line.getName())))
+              .collect(Collectors.toList());
+
+      // allow insert only if missing fields are nullable
+      if (checkNullabilityForFields(missingFields)) {
+        throw new RuntimeException("Non nullable fields are not allowed without schema.");
+      }
+
+      List<SchemaUtil.FieldWithIndex> tableFilteredFields =
+          tableSchema.getFields().stream()
+              .map(
+                  (tableField) -> {
+                    Optional<Schema.Field> optionalSchemaField =
+                        schema.getFields().stream()
+                            .filter((f) -> SchemaUtil.compareSchemaField(tableField, f))
+                            .findFirst();
+                    return optionalSchemaField.map(field -> FieldWithIndex.of(
+                        tableField, schema.getFields().indexOf(field))).orElse(null);
+                  })
+              .filter(Objects::nonNull)
+              .collect(Collectors.toList());
+
+      if (tableFilteredFields.size() != schema.getFieldCount()) {
+        throw new RuntimeException("Provided schema doesn't match with database schema.");
+      }
+
+      return tableFilteredFields;
+    }
+
+    /**
+     * A {@link org.apache.beam.sdk.io.jdbc.JdbcIO.PreparedStatementSetter} implementation that
+     * calls related setters on prepared statement.
+     */
+    private class AutoGeneratedPreparedStatementSetter implements PreparedStatementSetter<T> {
```

Comment:
That really auto generated class?

Reply:
No, this class automatically generates a preparedStatementSetter.

Reply:
So maybe rename it in something like PreparedStatementSetterGenerator?

Reply:
This is not a generator, Since its instance is already a "PreparedStatementSetter" but not an object that generates "PreparedStatementSetter".